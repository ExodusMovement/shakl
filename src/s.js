import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
