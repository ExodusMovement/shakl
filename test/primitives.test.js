import React from 'react'

import { create as r } from 'react-test-renderer'

import { Text, TouchableOpacity, View } from 'react-native'

import s from '../src/rn'

it('exposes primitives', () => {
  expect(s.Text).toBeDefined()
  expect(s.Touchable).toBeDefined()
  expect(s.View).toBeDefined()
})

test('exposed primitives are valid', () => {
  const Foo = s(Text)()
  const Bar = s(TouchableOpacity, { displayName: 'StyledTouchable' })()
  const Baz = s(View, { displayName: 'StyledView' })()

  const FooPrimitive = s.Text()
  const BarPrimitive = s.Touchable()
  const BazPrimitive = s.View()

  const foo = r(<Foo />).toJSON()
  const bar = r(<Bar />).toJSON()
  const baz = r(<Baz />).toJSON()

  const fooPrimitive = r(<FooPrimitive />).toJSON()
  const barPrimitive = r(<BarPrimitive />).toJSON()
  const bazPrimitive = r(<BazPrimitive />).toJSON()

  expect(JSON.stringify(foo)).toEqual(JSON.stringify(fooPrimitive))
  expect(JSON.stringify(bar)).toEqual(JSON.stringify(barPrimitive))
  expect(JSON.stringify(baz)).toEqual(JSON.stringify(bazPrimitive))

  expect(foo).toMatchSnapshot()
  expect(bar).toMatchSnapshot()
  expect(baz).toMatchSnapshot()
  expect(fooPrimitive).toMatchSnapshot()
  expect(barPrimitive).toMatchSnapshot()
  expect(bazPrimitive).toMatchSnapshot()
})
