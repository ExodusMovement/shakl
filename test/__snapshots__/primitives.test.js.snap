// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exposed primitives are valid 1`] = `
<Text
  style={{}}
/>
`;

exports[`exposed primitives are valid 2`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
/>
`;

exports[`exposed primitives are valid 3`] = `
<View
  style={{}}
/>
`;

exports[`exposed primitives are valid 4`] = `
<Text
  style={{}}
/>
`;

exports[`exposed primitives are valid 5`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
/>
`;

exports[`exposed primitives are valid 6`] = `
<View
  style={{}}
/>
`;
